### DISCUSSION TITLE
Companion App - Incorrect deltas / trend arrows and issues with the time it syncs

### DISCUSSION THREAD

**Discussion by petalina** [NightscoutFoundation/xDrip#4670] [Bug reports]
### Discussed in https://github.com/NightscoutFoundation/xDrip/discussions/3784

<div type='discussions-op-text'>

<sup>Originally posted by **petalina** November 26, 2024</sup>
Hello, since I switched from Dexcom One to Dexcom One+, using the Companion App as source, I have some issues with the arrows / difference of values and the time of update. I think somehow it is stuck in the time the older sensor used to update the values or something (eg some of the time points below - 1st pic: 8.47 / 103 mg/dl - 8.51 / 103 mg/dl - 8.56 / 103 mg/dl - 8.57 / 113 mg/dl, but delta: +41 mg/dl instead of 10 mg/dl) but I can't understand why the deltas it displays are so large even though the diagram and the values are correct. Also, at the same time the difference may be correct in the widget but with the wrong arrow (based on the real delta and xDrip's settings, not compared to the Dexcom app) and when I open the app both the arrow and difference are wrong. 
I installed the latest version of xDrip+ yesterday and also tried to Restart Collector and restart my phone.

![Screenshot 2024-11-26 162155](https://github.com/user-attachments/assets/f2d94112-a6f4-4ca9-bd7e-a27cec0d57fb)
![Screenshot 2024-11-26 163223](https://github.com/user-attachments/assets/678a3092-1cc5-4e76-ad6f-90d6dd4a3e99)</div>

**Comment by Qxlkdr** [comment]
I can agree. I'm using xDrip+ since 2 days ago and loves it. Only using it as follower using my kids credentials to follow his Dexcom share (Dex Share).

Dexcom Follow is showing a "Falling slightly" arrow but xDrip+ is showing a straight down arrow:
![Screenshot_20241209-010810~2](https://github.com/user-attachments/assets/344b379c-7bc3-4759-8e32-4450ae4cf56b)

At the same time, this is the result from the Home Assistant integration:
![Screenshot_20241209-010842](https://github.com/user-attachments/assets/3e3572e0-522a-415a-8dbe-1903344f4bd2)

**Comment by overfrenk** [comment]
Same problem with eversense in companion mode, without companion mode the problem disappears! I could not solve the problem in any way.

**Comment by Navid200** [comment]
@Qxlkdr @overfrenk 
The reason this issue was opened is the discrepancy between the difference between the last two xDrip readings and the xDrip delta value of 41 in the very first image in this thread.
If you look at that image, you can see that the difference between the last two readings is less than 41.  

However, xDrip is not supposed to show the delta or the trend arrow that Dexcom app or any other app shows.  xDrip is supposed to calculate the delta from the current and previous reading and calculate the arrow from that as well.
Dexcom app does not do that and uses a filer to come up with the arrow xDrip does not.  We are not trying to present Dexcom app's arrow in xDrip.  

If that is what you are interested in, you can open a discussion and we can talk about it.  
Any other post in this issue that is not about the real issue it represents may be ignored.

**Comment by overfrenk** [comment]
The problem is the same, regardless of the BG source, just use companion mode! If you notice you will have several readings with intervals of 1 or 2 minutes instead of the classic 5 minutes. I also obviously have the trend arrow that shows badly and hallucinatory delta values ​​even at +/-50.

**Comment by Navid200** [comment]
@overfrenk Why would the timing impact delta?  You would be right if xDrip showed the slope.  But, it does not.  It shows delta, which has nothing to do with the time axis.  Why would we get a delta of 40 instead of 10, as explained clearly in the very first post of this issue, just because the readings arrive at a timing other than once every 5 minutes?

**Comment by petalina** [comment]
This observation about the time interval and the incorrect delta seems to be correct and it was a thought I quote from the initial discussion:
"Is there any chance that it uses the +10 md/dl delta in 1 minute (according to the time points but not the reality / Dexcom app) as 10 mg/dl/min and converts it to an "expected" +41 mg/dl in 4 extra minutes?"

The incorrect deltas are proportional to the 1 min. delta times ~4.

**Comment by Navid200** [comment]
@petalina Please forget the arrows for now.  
xDrip calculates the arrows differently than Dexcom app.  We have no plan to change that.

I asked this issue to be opened because I am not sure why there is an error in the delta.
If we address the delta, I will consider this closed.  

With respect to the delta, how long does the error last?
I mean do you get this error for a few readings after you start a new sensor and then it goes away?

**Comment by petalina** [comment]
@Navid200 I didn't refer to the arrows in the last message but the incorrect deltas. The problem persisted for all the days the sensor lasted (10 days) though I changed back to Dexcom one to check what happens and there was no problem, the deltas are correct and a new value / point in the diagram is present every 5 minutes. 

When I use the Dexcom one+, after a couple of times having a new value every 5 minutes, I get a new one 1 minute after the last; that's the pattern. Also, whether the last value is 1 or 5 minutes apart from the previous, the widget always has the right delta and simultaneously the app and the notification the incorrect. But the widget, while having the correct delta has incorrect arrow according to this delta and xDrip's settings (not considering the Dexcom app). If the interval between the values is 5 minutes, there's no issue with the app / notification as well, so maybe the time interval plays some role.

**Comment by overfrenk** [comment]
The "only in companion mode" problem of the delta and the arrow only happens with the readings strangely very close, normally they should be every 4/5 minutes but for some strange reason they randomly arrive at 1/2 minutes. Example with red circles. I repeat if I use the same "eversense" sensor with readings via xdrip and esel the problem disappears.
![screen1](https://github.com/user-attachments/assets/1ed5ba7b-487d-4a02-9394-58ef75e28089)

**Comment by RansomTime** [comment]
This is affecting me. I think the observation by petalina is matching the behaviour I'm seeing.

Companinon app got a reading at 17:33 of 13.5 and then 17:34 it picked up the updated reading of 12.9, 

Didn't get a screenshot of the delta, but it was

I think this is stemming from the issue of it not always picking up the change until some time later.

For example, here is a reading on the pump app of 12.1 at 14:18, but xdrip is still showing 11.2

![Image](https://github.com/user-attachments/assets/9ef94e47-3490-4494-99a9-35815542f10c)

![Image](https://github.com/user-attachments/assets/e4b477a0-b932-4bc8-b145-e4c42e8d1d6a)

and again there are several times where there's 2 identical readings on the medtronic, and odd spacing which implies these aren't 5 minute apart readings.

![Image](https://github.com/user-attachments/assets/112b4dfb-df7e-47c9-802b-d7ecc3515fd5)

**Comment by overfrenk** [comment]
Same problem with eversense. It's a companion mode issue.

**Comment by adayper** [comment]
I'm having the same problem with the delay in Xdrip+ reading the notification in "companion app" mode. Most of the time, the value updates about 30 seconds before the next Carelink value changes. Since Carelink changes every 5 minutes, Xdrip+ has a delay of about 4 minutes and 30 seconds.

I don't think this happened a few months ago, or at least I didn't notice.

I've tried reinstalling everything and reconfiguring it, and it still happens. Sometimes it updates correctly, right on time, but most of the time it doesn't.

**Comment by Grntrenchman** [comment]
I am also getting this issue. The problem is that xDrip reads the data from the notification, but the data may be stale. Since readings are only accurate right when they happen, this causes some dissonance in the readings and delta. 

As an example: at 3:31 and 3:36 xdrip reads 82.7 mg/dl, but both are from the same sensor reading. then at 3:40 it picks up 79.0, which was actually taken minutes ago. then at 3:41 it picks up the new reading much earlier at 74.0. Because of the 1 minute gap between these readings, it produces a large probable delta, calculating it as a drop of 5 mg/dl in one minute. 

It fixes itself on the next read usually but this can cause confusion and false positives of low/high predictions and fall/climb rates, not to mention affecting suggested insulin/carb intake.

At least for G7, the Dexcom app doesn't give time of reading or delta in the notification, so I'm not sure how this can be fixed programmatically, except maybe some increased scanning in the beginning to find where the actual update happens and then only accepting one read per 5 minutes aligned with that.

<img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/e61048ce-0ca5-44fc-beb6-1e47ba75976a" />
<img width="2400" height="1080" alt="Image" src="https://github.com/user-attachments/assets/ccd6cd94-b0de-41af-841c-be30e1ee2be0" />

**Reply by Grntrenchman** [reply]
For other setups that cannot use Companion Bluetooth option: The easiest way to fix this would probably be to drop the preceding reading from calculations if a new reading comes in within a minute or two. 

The readings being so close due to xDrip not pulling the new reading exactly when it is made is what is causing all the issues here, on top of sometimes pulling the same reading twice.

I'm guessing that some/all of the apps don't update the notification right to the second, creating a slip in timing that eventually causes xDrip to pull a new reading near the end of it's 5-minute period, followed immediately by the new reading, causing xDrip to think the change is faster than it is.

A better way would be to, when 2 readings pile up (arrive in rapid succession), read back the few readings until you hit a duplicate, remove the leading duplicate, then space the readings in between the remaining duplicate and the 2nd reading of the piled up set. This would probably be much cleaner than just dropping the first of the 2nd piled readings.

<img width="793" height="367" alt="Untitled" src="https://github.com/user-attachments/assets/c4db1f73-38d8-47af-be16-4ff1539d56b1" />

**Comment by Navid200** [comment]
@Grntrenchman If you are using the companion app mode with a Dexcom device and experience the problem described here, it is 99% likely you have not enabled Companion Bluetooth.  You can see where it is here:
https://navid200.github.io/xDrip/docs/Follow/CompanionApp.html

**Reply by Grntrenchman** [reply]
Well I was going to say something about unwarranted personal attacks and speculation, but it's gone!

> If you are using the companion app mode with a Dexcom device and experience the problem described here, it is 99% likely you have not enabled Companion Bluetooth.

I have enabled this option and the readings immediately evened out.

<img width="1080" height="2400" alt="Screenshot_20260804-174448_xDrip+" src="https://github.com/user-attachments/assets/12b98eb1-e55a-4b8c-a9f8-b1afa83ea767" />

**Reply by Navid200** [reply]
Great!

Of course, the companion app mode is not as simple as one may think.  The Companion Bluetooth works with some sensors.  The behavior is also not going to be identical on all handsets.   But, since you had mentioned you used a Pixel, it would make it the most likely to succeed.

**Reply by Grntrenchman** [reply]
I would guess that vendor-specific android versions will be the least likely to work. This is on a Oneplus 7T, but with a lineage-based A16 rom, so it's very close to what a Pixel would be running.
